vllm/compilation/backends.py

```diff
@@ -164,106 +166,285 @@ def wrap_inductor(graph, example_inputs, additional_inductor_config):
+                    # cannot setattr to a module, so we need to set
+                    # the attribute in the __dict__
+                    self.split_gm.__dict__[
+                        item.submod_name] = PiecewiseBackend(
+                            item.graph, self.compilation_configs,
+                            self.graph_pool, is_first_graph)
+                    is_first_graph = False
+            returned_callable = self.split_gm
+
+        self.returned_callable = returned_callable
+        # trigger the first compilation
+        # TODO: if we can get the example inputs for each subgraph,
+        # we don't need to call it here
+        self.returned_callable(*example_inputs)
+
+        self._called = True
+
+        return self.returned_callable
+
+
+@dataclasses.dataclass
+class ConcreteSizeEntry:
+    runtime_shape: int
+    need_to_compile: bool  # the size is in compile_sizes
+    use_cudagraph: bool  # the size is in capture_sizes
+
+    compiled: bool = False
+    runnable: Callable = None  # type: ignore
+    num_finished_warmup: int = 0
+    cudagraph: Optional[torch.cuda.CUDAGraph] = None
+    output: Optional[Any] = None
+
+
+class PiecewiseBackend:
+
+    def __init__(self,
+                 graph: fx.GraphModule,
+                 compilation_configs: CompilationConfig,
+                 graph_pool: Any,
+                 is_first_graph: bool = False):
+        """
+        The backend for piecewise compilation.
+        It mainly handles the compilation and cudagraph capturing.
+
+        We will compile `self.graph` once for the general shape,
+        and then compile for different shapes specified in
+        `compilation_configs.compile_sizes`.
+
+        Independently, we will capture cudagraph for different shapes.
+
+        If a shape needs both compilation and cudagraph, we will
+        compile it first, and then capture cudagraph.
+        """
+        self.graph = graph
+        self.compilation_configs = compilation_configs
+        self.graph_pool = graph_pool
+        self.is_first_graph = is_first_graph
+
+        self.compile_sizes: Set[int] = set(
+            self.compilation_configs.compile_sizes)
+        self.capture_sizes: Set[int] = set(
+            self.compilation_configs.capture_sizes)
+
+        self.compile_finished = False
+        self.first_run_finished = False
+
+        self.compiled_graph_for_general_shape: Callable = None  # type: ignore
+
+        self.sym_shape_indices: List[int] = []
+
+        # the entries for different shapes that we need to either
+        # compile or capture cudagraph
+        self.concrete_size_entries: Dict[int, ConcreteSizeEntry] = {}
+        for shape in self.compile_sizes.union(self.capture_sizes):
+            self.concrete_size_entries[shape] = ConcreteSizeEntry(
+                runtime_shape=shape,
+                need_to_compile=shape in self.compile_sizes,
+                use_cudagraph=shape in self.capture_sizes,
+            )
```

Comment:
Would it be simpler to just keep two lists of entries, separate for CUDA graphs and compiled sizes?

Reply:
I don't think so. cudagraph needs to use the graph compiled from inductor.